src/or/control.c

```diff
@@ -5794,88 +5794,459 @@ control_event_bandwidth_used(uint32_t n_read, uint32_t n_written)
+    if (circ && CIRCUIT_IS_ORCIRC(circ)) {
+      return TO_OR_CIRCUIT(circ);
+    } else {
+      return NULL;
     }
+  }
+}
 
-    send_control_event(EVENT_PRIVCOUNT_DNS_RESOLVED,
-                       "650 PRIVCOUNT_DNS_RESOLVED %" PRIu64 " %" PRIu32 " %s\r\n",
-                       oncirc->p_chan->global_identifier,
-                       oncirc->p_circ_id,
-                       exitconn->base_.address);
+/* Like privcount_get_or_circuit, but with casts that tell the compiler we're
+ * not actually modifying anything */
+static const or_circuit_t*
+privcount_get_const_or_circuit(const edge_connection_t* exitconn,
+                               const or_circuit_t *orcirc)
+{
+  return (const or_circuit_t *)privcount_get_or_circuit(
+                                                (edge_connection_t *)exitconn,
+                                                (or_circuit_t *)orcirc);
 }
 
-void control_event_privcount_stream_data_xferred(edge_connection_t *conn, uint64_t amt, int is_outbound) {
-    if(!get_options()->EnablePrivCount || !EVENT_IS_INTERESTING(EVENT_PRIVCOUNT_STREAM_BYTES_TRANSFERRED)) {
-        return;
+/* Is conn an EXIT conn (or BEGINDIR conn), and does it end at this relay? */
+static int
+privcount_is_exit(const edge_connection_t* exitconn,
+                  const or_circuit_t *orcirc)
+{
+  if (!exitconn) {
+    /* Assume missing connections are *not* Exit connections */
+    return 0;
+  }
+
+  const or_circuit_t* oc = privcount_get_const_or_circuit(exitconn, orcirc);
+
+  if (!oc) {
+    /* Assume missing circuits are *not* OR circuits */
+    return 0;
+  }
+
+  /* if the circuit started here, this is our own stream and we can ignore it
+   */
+  if (CIRCUIT_IS_ORIGIN(TO_CIRCUIT(oc))) {
+    return 0;
+  }
+
+  if (!oc->p_chan) {
+    return 0;
+  }
+
+  return TO_CONN(exitconn)->type == CONN_TYPE_EXIT;
+}
+
+/* Should PrivCount ignore events from this connection or circuit?
+ * Directory connections and DNS resolves are considered overhead, so this
+ * function is not suitabe for filtering EVENT_PRIVCOUNT_DNS_RESOLVED events.
+ */
+static int
+privcount_is_overhead(const connection_t* conn, const circuit_t *circ)
+{
+#if PRIVCOUNT_DNS_RESOLVE_IS_OVERHEAD
+  if (privcount_conn_is_dns_resolve(conn)) {
+    return 1;
+  }
+#endif
+
+  /* if the circuit started here, this is our own stream and we can ignore it
+   */
+  if (circ && CIRCUIT_IS_ORIGIN(circ)) {
+    return 1;
+  }
+
+  return privcount_is_dir(conn, circ);
+}
+
+/* Should PrivCount count streams and bytes from this connection or circuit?
+ */
+int
+privcount_is_counted_for_bytes(const edge_connection_t* exitconn,
+                               const or_circuit_t *orcirc)
+{
+  /* Ignore events that are disabled */
+  if (!get_options()->EnablePrivCount) {
+    return 0;
+  }
+
+  if (!EVENT_IS_INTERESTING(EVENT_PRIVCOUNT_STREAM_BYTES_TRANSFERRED) &&
+      !EVENT_IS_INTERESTING(EVENT_PRIVCOUNT_STREAM_ENDED)) {
```

Comment:
I assume PrivCount would complain about the config if asked to count stream bytes but not streams, or otherwise would automatically enable both event types.